google/cloud/storage/async_object_requests.h

```diff
@@ -31,15 +32,21 @@ GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
 
 /**
  * An opaque representation of the data for an object payload.
- *
- * While applications do not need to create instances of this class, they may
- * need to use it in their mocks, to validate the contents of their requests.
  */
 class WritePayload {
  public:
   /// Creates an empty payload.
   WritePayload() = default;
 
+  /// Creates a payload from @p p.
+  explicit WritePayload(std::string p)
+      : impl_(storage_internal::MakeCord(std::move(p))) {}
+
+  /// Creates a payload from @p p.
+  template <typename T>
```

Comment:
Generic question for public APIs with templates....

So only certain `T`s work. Is there a better way to communicate which ones work than compiler errors for an internal API when they don't?

Should we expose `IsPayloadType`?

Reply:
> Is there a better way to communicate which ones work than compiler errors for an internal API when they don't?

In C++ < 20.  Basically no.  The alternatives are: named requirements in the documentation, and sometimes tricks with `std::enable_if<>`. With the latter the error messages are not guaranteed to be readable (though you can make them readable if you name the `std::enable_if<>` types cleverly.

In C++ >= 20: [constraints and concepts](https://en.cppreference.com/w/cpp/language/constraints)

> Should we expose `IsPayloadType`?

I don't think that would work, since this can handle `std::vector<std::byte>` and `std::vector<std::vector<std::byte>>`, and `std::vector<std::string>`.

We would need a concept or constraint to express: "A vector of things that look like bytes, or a vector of vectors of things that look like bytes". It may be easier to document what these do if we break them into multiple overloads.  

But in the future  it might be possible to support any container type that owns the underlying buffer.  For example, `std::list<std::vector<char>>` would not be hard to implement. 

"Efficiently convertible to `absl::Cord`" is kind of what we want to express, but that is not easy to translate that into code.

I am going to punt on this for now.

Reply:
Solving the problem with documentation works for me. Thanks.